---
title: "randomized groups"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
```

```{r load_libraries, include = FALSE}
# Load libraries
library(DESeq2)    # BiocManager::install('DESeq2')
library(limma)     # BiocManager::install('limma')
library(stringr)
library(readxl)
library(pheatmap)
library(RColorBrewer)
library(variancePartition)  # BiocManager::install('variancePartition')
library(doParallel)
library(cowplot)
library(tidyverse)
library(KOGMWU)  # install.packages("KOGMWU")
library(adegenet)
library(ggpubr)
library(furrr)

## ggplot theme
theme_custom <- function() {
  theme_bw(base_size = 10) %+replace%    #, base_family = "Arial"
    theme(
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(), 
      panel.background = element_blank(),
      panel.border = element_rect(color = "black", fill = NA),
      legend.background = element_rect(fill = NA, colour = NA),
      axis.text.x = element_text(angle=45, hjust=1, vjust = 1)#,
      #legend.title = element_text(size = 8), 
      #legend.text = element_text(size = 7)
    )
}
## ggplot labeller
colnames <- c(
  `20` = "genet 1",
  `22` = "genet 2",
  `26` = "genet 3"
)

contrastnames <- c(
  `Dc.Cc` = expression("D"[C]~"vs."~"C"[C]),
  `Dh.Dc` = expression("D"[H]~"vs."~"D"[C]),
  `Ch.Cc` = expression("C"[H]~"vs."~"C"[C])
)

global_labeller <- labeller(
  colony = colnames,
  contrast = contrastnames,
  .default = "label_parsed"
)
```

# Load data

##### Gene count data
```{r import_counts}
# Import gene expression / count data
counts <- read.table("data/tagseq/processed/counts_newpars.txt", 
                     header = TRUE, row.names = 1)
# Trim column names to just the sample name
colnames(counts) <- str_sub(colnames(counts), 1, 7)
# Order columns by sample name
counts <- counts[, order(colnames(counts))]
```

##### Sample metadata
```{r sample_metadata}
# Import sample metadata
sdat <- read_xlsx("data/sample_metadata.xlsx") %>%
  mutate(sym = recode(trt1, b = "D", c = "C"),
         sample = paste0(species, colony, ".", core),
         group = paste(sym, trt2, sep = ""),
         colony = factor(colony),
         colony.group = interaction(colony, group)) %>%
  mutate_if(is.character, as.factor) %>%
  arrange(sample) %>%                              # order rows by sample name
  column_to_rownames(var = "sample")               # set sample name to rownames
```

#### Create DESeqDataSet
```{r subset_dds}
# Create full DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = sdat,
                              design = ~ colony)

# Subset DESeqDataSet
## Subset time point at end of heat stress
dds <- dds[, colData(dds)$date_sampled == "2017-10-28"]   
### Remove genes counted zero times in subset
dds <- dds[ rowSums(counts(dds)) > 0, ]

## Subset only the colonies in the heat stress experiment
hs.dds <- dds[, colData(dds)$colony %in% c(20, 22, 26)]   
### Remove genes counted zero times in subset
hs.dds <- hs.dds[ rowSums(counts(hs.dds)) > 0, ]
### Drop unused factor levels
colData(hs.dds) <- droplevels(colData(hs.dds))

# Set design formula to test differences among groups accounting for colony
design(hs.dds) <- formula(~ colony + group)

# Filter out sample that did not shuffle -- Mc22.12 (17th sample in hs.dds)
hs.dds <- hs.dds[, -17]

# Remove genes with less than 1 mean count across samples
hs.dds <- hs.dds[rowMeans(counts(hs.dds)) > 1, ]
```

#### RANDOMIZE GROUPING VARIABLES -- symbiont and heat stress treatment, within each colony
```{r}
actual_colData <- colData(hs.dds)
# RANDOMIZE SYMBIONTS and heat stress WITHIN EACH COLONY
nsim <- 1500

# Create n=nsim randomized colData Data Frames for DESeqDataSet
rand_colData <- list()
for (i in 1:nsim) {
  rand_colData[i] <- data.frame(colData(hs.dds)) %>%
  group_by(colony, trt2) %>%
  mutate(sym = base::sample(sym)) %>%
  group_by(colony, sym) %>%
  mutate(trt2 = base::sample(trt2),
         group = paste0(sym, trt2),
         colony.group = paste(colony, group, sep=".")) %>%
  ungroup() %>%
  mutate(group = as.factor(group),
         colony.group = as.factor(colony.group)) %>%
  DataFrame()
}
```

```{r}
# Set up for parallel processing of randomized datasets in downstream analyses
plan(multiprocess)
```

# Differential expression (DE) analysis

Since colonies appear to respond differently to symbiont type and heating, we should analyze differential expression in each colony separately to assess their unique responses. We can also analyze differential expression across all colonies together to see what responses are common to all three.


##### Run DESeq analysis for all genets together

```{r run_DESeq_colonies, warning=FALSE}
# Define group contrasts - Need to sum DEGs from all possible pairwise combinations to compare to random groupings
group.contrasts <- tibble(num = c("Dh", "Dh", "Dh", "Dc", "Dc", "Ch"),
                          den = c("Dc", "Ch", "Cc", "Ch", "Cc", "Cc"))

# Function to run DESeq for given version of colData, get DEGs and logP values
get_degs <- function(rd) {
  # Create DESeq object with random grouping data (rd)
  dds <- hs.dds
  colData(dds) <- rd
  # Run DESeq pipeline
  dsr <- DESeq(dds, minmu = 0.5)
  # Build result
  res <- tibble(dsr = list(dsr)) %>%
    expand_grid(group.contrasts) %>%
    # Get DESeq result for each contrast
    mutate(res = pmap(list(dsr, num, den), function(dsr, num, den) {
      results(dsr, contrast = c("group", num, den))})) %>%
    # Get number of DEGs for each contrast
    mutate(nsig = map_dbl(res, ~ nrow(data.frame(.[which(.$padj < 0.1), ])))) %>%
    # Get median log2FoldChange across all genes for each contrast (will filter on this later)
    mutate(medl2fc = map_dbl(res, ~ median(data.frame(.)$log2FoldChange))) %>%
    # Get logP values for downstream KOG analysis
    mutate(logP = map(res, ~ data.frame(
      gene = rownames(data.frame(.)),
      logP = -log10(data.frame(.)$pvalue) * sign(data.frame(.)$log2FoldChange)))) %>% 
    select(num, den, nsig, medl2fc, logP)
  # Return results
  return(res)
}

# Run DESeq and collect results for random simulations and actual data
random <- future_map(rand_colData, get_degs)
actual <- get_degs(actual_colData)

# Bind results together
res <- random %>% 
  bind_rows(.id = "sim") %>%
  bind_rows(random = ., actual = actual, .id = "colData")

# Summarize results
ressum <- res %>%
  group_by(colData, sim) %>%
  summarise(
    # Add up total DEGs from all 6 contrasts
    totsig = sum(nsig),
    # Of the 6 contrasts, find the highest deviation from zero in l2fc (tends to return inflated DEGs)
    maxmedl2fc = max(abs(medl2fc))) %>%
  arrange(as.numeric(sim))
```

```{r}
# Plot results

## Total DEGs from all simulations, plotted against maximum median log2foldchange
ggplot(ressum, aes(x = maxmedl2fc, y = totsig, color = colData)) + 
  geom_point()
# This plot shows that when ANY contrast among the 6 has an atypically high median log2foldchange across all genes, an abnormally high number of DEGs is produced. This is caused by chance when the number of read counts differs between the groups in the contrast. We throw out these cases because the high number of DEGs is caused by this random difference in read counts between the groups that prevents DESeq from effectively normalizing the counts. (see https://www.biostars.org/p/239253/)

# Filter out random simulations not passing threshold
ressum <- ressum %>%
  filter(maxmedl2fc < 0.5)

# Get actual number of DEGs
n.act <- ressum %>%
  filter(colData == "actual") %>%
  pluck("totsig")

# Calculate p-value (proportion of the random simulations that
#  pass maxl2fc threshold that return >= the actual number of DEGs)
deg.out <- ressum %>%
  filter(colData == "random") %>%
  summarise(p = sum(totsig >= n.act)/n(),
            rmed = median(totsig))

# Plot total number of DEGs for all cases that pass maxmedl2fc < 0.5 threshold
ggdeg <- ressum %>%
  filter(colData == "random") %>%
  ggplot(aes(x = totsig)) +
  geom_histogram() +
  geom_vline(xintercept = c(deg.out$rmed, n.act), lty = 2) +
  theme_custom() +
  annotate("text", x = deg.out$rmed+10, y = 275, label = paste0("Median: ", deg.out$rmed), hjust = 0, size = 2) +
  annotate("text", x = n.act+10, y = 275, label = paste0("Actual DEGs: ", n.act), hjust = 0, size = 2) +
  annotate("text", x = n.act+10, y = 255, label = paste0("p=", round(deg.out$p, 4)), hjust = 0, size = 2) +
  labs(x = "Total significant DEGs", y = "Number of random data permutations")
ggdeg
```

# EuKaryotic Orthologous Groups (KOG) enrichment analysis

Using the signed log(p-values) from the differential expression analysis, we can analyze which KOG classifications are up- or down-regulated within each colony, and across all three colonies together.

```{r KOG, results = 'hide', eval = FALSE}
# Import KOG annotations for Mcavernosa genome # Downloaded from M. Studivan github
gene2kog <- read.table("data/genome/Mcavernosa_iso2kog.tab", sep = "\t") %>% 
   filter(!V2 == "")

# Get only simulations that pass maxmedl2fc threshold
pass <- pluck(ressum, "sim")

# Run KOG.MWU analysis on all passing permutations
res <- res %>%
  filter(sim %in% pass) %>%
  mutate(KOG = future_map(logP, ~ kog.mwu(., gene2kog)))

# count significant kog categories
res <- res %>%
  mutate(nsigKOG = map_dbl(KOG, ~ sum(.$padj < 0.1)))

# add up sig kog categories for all contrasts in each simulation
kogsumm <- res %>%
  group_by(colData, sim) %>%
  summarise(totnsigKOG = sum(nsigKOG))

# Actual number of significant KOG categories
act.nkog <- kogsumm %>%
  filter(colData == "actual") %>%
  pluck("totnsigKOG")

# Median significant KOGs from random, and p-value
kog.out <- kogsumm %>%
  filter(colData == "random") %>%
  summarise(p = sum(totnsigKOG >= act.nkog)/n(),
            rmed = median(totnsigKOG))

# Plot number of sig kog categories
ggkog <- kogsumm %>%
  filter(colData == "random") %>%
  ggplot(aes(x = totnsigKOG)) + 
  geom_histogram() +
  theme_custom() +
  geom_vline(xintercept = c(kog.out$rmed, act.nkog), lty = 2) +
  annotate("text", x = kog.out$rmed+0.5, y = 130, label = paste0("Median: ", kog.out$rmed), hjust = 0, size = 2) +
  annotate("text", x = act.nkog-0.5, y = 130, label = paste0("Actual KOGs: ", act.nkog), hjust = 1, size = 2) +
  annotate("text", x = act.nkog-0.5, y = 120, label = paste0("p=", round(kog.out$p, 4)), hjust = 1, size = 2) +
  labs(x = "Total significant KOGs", y = "Number of random data permutations")
ggkog
```

```{r}
ggout <- cowplot::plot_grid(ggdeg, ggkog, labels = "auto")
ggsave(filename = "figures/FigS10.png", width = 169, height = 80, units = "mm")
```

